Implement tag CRUD

jessie-ross opened the pull request:
<!-- Describe your changes in detail. -->
**This PR implements Tags CRUD.**

This is a working copy and paste job of Metrics, useful for testing the backend implementation of tags.

Video: https://d.pr/v/1KJC71

<!-- Remember to include context: Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->
<!-- Delete any bullet points that don't apply and add more details if needed. -->

- (Todo) Automated tests that cover added/changed functionality
- Manual testing with production data (locally)
- (Todo) Manual testing with mock data (locally or on Vercel)

jessie-ross commented:
Still doesn't look like we are in refactoring town for the modals, but we are on the way there.

I would like to do some reorganising of `/components` though, in a separate PR after this, a lot of renaming and adding directories.

jessie-ross commented:
For the reviewer, it will be useful to perform diffs between these files and their Metric counterparts..

withinboredom commented:
> For the reviewer, it will be useful to perform diffs between these files and their Metric counterparts..

It would be nice if the first commit of the branch was just literally what you copied over. Then `git diff` or tools like `tig` could easily show what specifically changed since that commit. It would also be useful for future refactorings and bug hunts, but we squash commit our PR's so it actually won't be useful beyond the review ;) but just something to point out for future reference.

jessie-ross commented:
> It would be nice if the first commit of the branch was just literally what you copied over.

Ah that's a good idea!

withinboredom reviewed (commented):
Still reviewing, it's a bit of code. But

jessie-ross commented:
Maybe hold off on a detailed review for the moment, this may break after adding "experiment_ids" to Tags - I am going to have to split the Tag schema into TagBare and TagFull.

jessie-ross commented on src/components/TagsTable.tsx:
```diff
@@ -0,0 +1,65 @@
+ * @param tags An array of tags.
+ * @param onEditTag A Callback. Setting this will show the edit action in the table.
+ */
+const TagsTable = ({ tags, onEditTag }: { tags: Tag[]; onEditTag?: (tagId: number) => void }): JSX.Element => {
```
Not sure I follow this, this is for the tags CRUD where we want to display a table of tags.

We will be using chips for input/display of tags on experiment if that's what you mean?

withinboredom replied:
Ah, yeah, that's what I meant :)

jessie-ross commented:
> Maybe hold off on a detailed review for the moment, this may break after adding "experiment_ids" to Tags - I am going to have to split the Tag schema into TagBare and TagFull.

I have decided to split it anyway to keep it consistent and at the same time not include `experiment_ids`, see my latest discussion on the diff.

This is ready for review.

aaronyan commented:
> I have decided to split it anyway to keep it consistent and at the same time not include experiment_ids, see my latest discussion on the diff.

Moving the discussion [over here](https://betterexperiments.wordpress.com/2020/10/05/experiment-tags-as-enablers-of-many-features/#comment-1239) so it's all in one place!

jessie-ross commented:
Thanks for moving it to the P2 Aaron, that makes a lot of sense.

Either way `experiment_id` goes we can alter this code in the future to suite, especially as I have split tags into TagBare and TagFull. So this is ready to review.

jessie-ross commented:
I might actually redo this PR with @withinboredom's suggestion of copying the files, commiting and then changing, to make this easier to review.

jessie-ross commented:
Ok, hopefully I have made this a bit easier to review:
- I have changed the base of this PR to another PR that is just a copy of the metrics files.
- I have reverted the Improve schema types PR that accidentally got included in here.

withinboredom reviewed (approved):
So much easier to review! Thanks for changing the base like that, it's clear what's new vs. changed from somewhere else. At some point, we probably want to combine these things into something more general, but I don't think that point is right now.

:ship: it
